### ISSUE TITLE
Support for 'and' as part of a command

### ISSUE THREAD

**Issue by ecoble** [uchicago-cs/chiventure#275]
Support the 'and' keyword as part of user input so that two commands can be processed at the same time. e.g. "Take orb and walk east" This will largely involve adjustments to the parser

**Comment by mberamendiconde** [comment]
Some thoughts on this issue firsthand:
i can think of a way to approach this:
we don't have to change the TOKEN_LIST_SIZE variable value,
instead we can find a way to ensure that the implementation can recognize "and" and rather than adding that word to the words array, we start a NEW words array (recall the function with the rest of the input) (ie there are 2 word arrays total)

**Comment by Beza20** [comment]
Had a zoom chat with Martha about the issue. A potential solution we are thinking of is just to add "and" to the array of delimiters. But we are concerned about the "and" always coming after a space. e.g. if we have 2 commands x,y and then the user input command as such "xandy", we are not sure how we want our function to handle that scenario. Ideally, it shouldn't matter but we need to find out more about the capabilities of strtok. Planning for another meeting with Elizabeth to discuss issue.

**Comment by Gestalt-R** [comment]
Might as well leave in my two bits. For some reason, I've had this issue assigned for two years, seeing as how I and Elizabeth designed the original tokenizer. strtok can totally be used with multiple delimiters, but they have to be _characters_ rather than strings. You could dodge this by using a for loop (with multiple separate instances of strtok), but that also gets tricky. One approach might be to split the string before tokenizing it: interpreting "x and y" as the lines "x" and "y."

And with that, I'll disable notifications on this issue! Best of luck guys!

**Comment by mberamendiconde** [comment]
So I tried an approach with strstr by which if the " AND " substring is found in the input, I used strncpy to replace it with " ". That way, the code already in place with the loop for strtok to tokenize based on spaces could still be used, and we can avoid having an "and" token. I'm not entirely sure if what I did was correct BUT i used a while loop to make sure that more than one and could be changed to a " "

**Comment by Beza20** [comment]
A course of action that I would like to try to solve this problem is to implement a helper function: strtokstr() that essentially functions exactly as the built in C function strtok_r does and takes in the same parameters except that it takes in a string as a delimiter instead of a character which was what we failed to realize for our first proposed solution. This helper function will be convenient since it doesn't limit a delimiting string to "and" and opens up other possibilities for a user to input multiple commands using different connecting words. After implementing this helper function, it will be called inside the parse function in a while loop where through each iteration it will store into the token array the word that was on the left side of the delimiter string and removes from the string the added word and the delimiter to implement the function on the rest of the string.

**Comment by Gestalt-R** [comment]
Ok, I said I didn't want to comment more, but I somehow still got a notification.

This is not how tokens are processed in the code as is. A token is a singular verb or noun. For example `open door` passes two tokens. Delimiters separate these words. Making "and" a delimiter doesn't allow you to process a phrase `take milk and take cookies` as `take milk` and `take cookies`, but rather `take milk cookies`. You can take the other approach, but you will need to rewrite how tokens are handled, period, which should involve a massive refactoring of the code, if admittedly a necessary one. (Even if you allow dynamic lists of tokens, you will need to include "and" as one of them, in order to separate commands, or perhaps make a parser intelligent enough to know which tokens belong with which commands.) 

Another approach would be to preprocess the input, for example, treating certain strings as newline characters.

**Comment by Beza20** [comment]
But we had that problem even when we were tokenizing by space only. There is a way to handle the situation which is that once we have a tokenized list of words not all of those words are going to be "commands." Yes, we should make that list more flexible and I can assure you the  CLI team right now is working on that but the code as it is right now can take a list of words tokenized however way(character or string(once we get that done)) and operate accordingly. the different kinds of operations are what they're there for. If I am missing something please do let me know.

**Comment by Gestalt-R** [comment]
The thing I think you're missing is in the code, in its current state, the tokens are not necessarily commands, but rather, commands OR their arguments. An array of tokens is then sent to the parser to process them into calls of operations with the remaining tokens as the arguments of the operations. Each array is a single call of an operation. Of course, maybe this is all just a failure to understand, on my end.

Let's walk through a few testcases of the code. I type `take cake pie cookie sausage` into the console, and this calls `cmd_from_string` in the cmd.module on that input. The parser module attempts to tokenize the input. It does a few iterations, setting the first token as `take`, the second as `cake`, and so on. The for loop terminates before tokenizing `sausage`, and so, outputs a NULL pointer, indicating an invalid input. 

I type `combine cake pie` into the console. The parser module attempts to tokenize the input. It does a few iterations, setting the first token as `combine`, the second as `cake`, and so on. The for loop terminates on an empty string, and outputs a pointer to an array of tokens: `combine`, `cake`, `pie`.

Then, `cmd_from_tokens` is called on that list of tokens. The first token, `combine`, is used to lookup the correct operation, which has already been added, either built-in, in which case it is explicitly defined in operations.c (like HELP), or procedurally, using add_cmd in cmd.c, in which case, it maps to an operation kind. In this case, `combine` is not built-in, so it maps to a `kind3_action_operation`. Note that commands that are not built-in need to reference their first token, corresponding to the command being used, as the operation takes place. 

Then, the operation, which has been defined already to be a KIND 3 operation while the program was booting up, is called. Functionally, this means it runs `kind3_action_operation(tokens, context_struct)`. This function applies a change of state to the context struct (basically, it makes the game do something), looking at the first token, `combine` to determine what it should do, and the other tokens to determine what items it should do the operation to.

I'm pretty puzzled as to what "that problem" is.  You seem to be under the impression that the code already views each token as a command, and are trying to make it tokenize all words, but it already is that way. Making the list more flexible is not difficult, at least to do so sloppily: (as said earlier, for testing purposes you can actually change the size by changing line 9 in parser.h, but you may want to add a couple extra measures for checking invalid input if so). And yes, your understanding of how operations work in the abstract is good, but you might want to check on how operations reference the tokens that they are given.

TL; DR 

The first approach of using "and" as a delimiter really doesn't make sense, either with or without the proposed refactoring/changes. The 2nd approach you mentioned, of handling a dynamic list of tokens, should work just fine, but requires _significant_ changes to the way the code currently works. (i.e, each operation has to "know" exactly how many tokens to expect in order to segment the input into sets of commands, meaning you need some sort of pre/post-parser) That being said, if you do use that, I really don't see the point of using the first method.

**Comment by Gestalt-R** [comment]
P.S

As the code is currently written, you would have to rewrite _everything_ to use a dynamic list if you decide to go down the route of having no maximal token list length, in addition to processing the dynamic list, which would probably involve at least twice as much work as the former. In contrast, preprocessing the input into multiple lists of tokens ("commands") is a relatively quick fix.

**Comment by Gestalt-R** [comment]
Oh, and one last thing: it seems like the purpose of this issue kinda shifted at some point from running two commands in a single line to running commands that happen to have the word "and" in them. You could probably just make a new operation type for propositional commands, and handle it that way without needing to deal with the parser at all.

**Comment by ecoble** [comment]
@Gestalt-R There might be some miscommunication here. What the current CLI team is doing is the original goal of the the issue, two process two commands from a single user input. Their design is that they will first split the input on the word "and" (or theoretically any "connector" word), giving them two separate inputs to parse. So the string "take milk and take cookies" will parse to one string "take milk" and another string "take cookies," each of which will then be separately handled by the existing tokenization code.

**Comment by Gestalt-R** [comment]
In that case, all my other comments should be read as directly addressing the issue they thought they were working on.

**Comment by mberamendiconde** [comment]
To document our progress on this issue: We have implemented a helper function called strtokstr_r , which solves this issue #275

We will submit a pull request for this branch (cli/parse). 

We also plan to create a separate issue which will expand on the capabilities of this helper function. However, for this issue, the parser can now deal with the "and" input, and thus the issue is solved.

**Comment by Beza20** [comment]
The key to this issue is to understand how a command line is read and then executed. After a parser function(breaking it up using a string and then a space) is applied to a cmd line, it is turned to a command struct so that an action type is assigned to it in the cmd module. But this function is actually called to do that in the ui module and it is called once. So in order to make sure that once a command line is read and parsed we want the function that changes a parsed string to a cmd struct and executes it to be computed until all of the commands connected by "and" are iterated through.

**Comment by Beza20** [comment]
we have enabled a user to input multiple commands at once using a connecting word "and." This was done by using a helper function <strtokstr_r> that has very similar structure as the C built-in function <strtok_r.> It parses through a string and tokenizes a segment of that string whenever it finds the inputed delimiter of a string type. Then we use the helper function to read the cmd line and store the tokenized segment of the input into a utlist. Note: Each element in the utlist has to further be parsed breaking it up using a space. And then in the UI module, we made sure that all of the command segments put into the utlist are iteratively changed into a command struct by calling the <cmd_from_string> and executed. And with that this pull request has introduced the "and" feature to chiventure resolving issue #275.

**Comment by ecoble** [comment]
Issue Score: ✔️++

Comments:
Excellent work!! This was a really tough issue that I know was frustrating at times, you did a really great job, and the documentation is excellent too!
